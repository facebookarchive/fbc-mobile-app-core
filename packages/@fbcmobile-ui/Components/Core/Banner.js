/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 */

'use strict';

import * as React from 'react';
import Text from '@fbcmobile/ui/Components/Core/Text';
import {Animated, Easing, StyleSheet} from 'react-native';
import {Colors} from '@fbcmobile/ui/Theme';
import {useEffect, useState} from 'react';

const BANNER_HEIGHT = 54;
const BANNER_DEFAULT_TIMEOUT_MS = 3000;

type Props = {
  +message: React.Node,
  +timeoutMs?: number,
  +skin?: 'dark' | 'green' | 'red',
};

const Banner = ({
  message,
  timeoutMs = BANNER_DEFAULT_TIMEOUT_MS,
  skin = 'dark',
}: Props) => {
  const [topValue] = useState(new Animated.Value(-BANNER_HEIGHT * 3));
  useEffect(() => {
    const timeoutId = setTimeout(() => {
      Animated.timing(topValue, {
        toValue: -BANNER_HEIGHT * 3,
        friction: 1,
        easing: Easing.ease,
        useNativeDriver: false,
      }).start();
    }, timeoutMs);
    return () => clearTimeout(timeoutId);
  });
  useEffect(() => {
    Animated.timing(topValue, {
      toValue: 0,
      friction: 1,
      easing: Easing.ease,
      useNativeDriver: false,
    }).start();
  }, [topValue]);

  return (
    <Animated.View
      style={[
        styles.root,
        styles[`${skin}Skin`],
        {
          top: topValue,
        },
      ]}>
      <Text
        variant="h6"
        color={skin === 'red' ? 'red' : 'light'}
        align="center">
        {message}
      </Text>
    </Animated.View>
  );
};

const styles = StyleSheet.create({
  root: {
    position: 'absolute',
    left: 0,
    right: 0,
    zIndex: 1000,
    paddingVertical: 17,
    paddingHorizontal: 16,
  },
  darkSkin: {
    backgroundColor: Colors.BlueGray,
  },
  greenSkin: {
    backgroundColor: Colors.Green,
  },
  redSkin: {
    backgroundColor: Colors.RedBackground,
  },
});

export default Banner;
