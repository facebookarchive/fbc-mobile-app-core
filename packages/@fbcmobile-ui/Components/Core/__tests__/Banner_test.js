/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 */

'use strict';

import Banner from '../Banner';
import React from 'react';
import renderer from 'react-test-renderer';

jest.mock('react-native/Libraries/Animated/src/Animated', () => {
  const ActualAnimated = jest.requireActual(
    'react-native/Libraries/Animated/src/Animated',
  );
  return {
    ...ActualAnimated,
    timing: (value, config) => {
      return {
        start: callback => {
          value.setValue(config.toValue);
          callback && callback();
        },
      };
    },
  };
});

describe('Banner Tests', () => {
  test('Banner renders correctly with message and timeout', () => {
    global.managedClock(() => {
      const tree = renderer.create(<Banner message={'hello'} timeoutMs={10} />);
      const data = tree.toJSON();
      global.moveTimeForward(10);
      expect(data).toMatchSnapshot();
    });
  });
  test('Banner renders correctly with message and skin', () => {
    global.managedClock(() => {
      const tree = renderer.create(<Banner message={'hello'} skin={'red'} />);
      const data = tree.toJSON();
      global.moveTimeForward(15);
      expect(data).toMatchSnapshot();
    });
  });
});
