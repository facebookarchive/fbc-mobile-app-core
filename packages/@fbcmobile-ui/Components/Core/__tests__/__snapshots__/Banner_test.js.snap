// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Banner Tests Banner renders correctly with message and skin 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#FFEAE6",
      "left": 0,
      "paddingHorizontal": 16,
      "paddingVertical": 17,
      "position": "absolute",
      "right": 0,
      "top": -162,
      "zIndex": 1000,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#484848",
          "fontSize": 15,
          "fontWeight": "500",
          "lineHeight": 20,
        },
        Object {
          "color": "#F02849",
        },
        null,
        Object {
          "textAlign": "center",
        },
      ]
    }
  >
    hello
  </Text>
</View>
`;

exports[`Banner Tests Banner renders correctly with message and timeout 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#484848",
      "left": 0,
      "paddingHorizontal": 16,
      "paddingVertical": 17,
      "position": "absolute",
      "right": 0,
      "top": -162,
      "zIndex": 1000,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#484848",
          "fontSize": 15,
          "fontWeight": "500",
          "lineHeight": 20,
        },
        Object {
          "color": "white",
        },
        null,
        Object {
          "textAlign": "center",
        },
      ]
    }
  >
    hello
  </Text>
</View>
`;
